modules/navier_stokes/src/linearfvkernels/LinearWCNSFVMomentumFlux.C

```diff
@@ -0,0 +1,358 @@
+  params.addRequiredParam<MooseFunctorName>(NS::mu, "The diffusion coefficient.");
+  MooseEnum momentum_component("x=0 y=1 z=2");
+  params.addRequiredParam<MooseEnum>(
+      "momentum_component",
+      momentum_component,
+      "The component of the momentum equation that this kernel applies to.");
+  params.addParam<bool>(
+      "use_nonorthogonal_correction",
+      true,
+      "If the nonorthogonal correction should be used when computing the normal gradient.");
+  params.addParam<bool>(
+      "use_deviatoric_terms", false, "If deviatoric terms in the stress terms need to be used.");
+
+  params += Moose::FV::advectedInterpolationParameter();
+  return params;
+}
+
+LinearWCNSFVMomentumFlux::LinearWCNSFVMomentumFlux(const InputParameters & params)
+  : LinearFVFluxKernel(params),
+    _dim(_subproblem.mesh().dimension()),
+    _mass_flux_provider(getUserObject<RhieChowMassFlux>("rhie_chow_user_object")),
+    _mu(getFunctor<Real>(getParam<MooseFunctorName>(NS::mu))),
+    _use_nonorthogonal_correction(getParam<bool>("use_nonorthogonal_correction")),
+    _use_deviatoric_terms(getParam<bool>("use_deviatoric_terms")),
+    _advected_interp_coeffs(std::make_pair<Real, Real>(0, 0)),
+    _face_mass_flux(0.0),
+    _stress_matrix_contribution(0.0),
+    _stress_rhs_contribution(0.0),
+    _index(getParam<MooseEnum>("momentum_component")),
+    _u_var(dynamic_cast<const MooseLinearVariableFVReal *>(
+        &_fe_problem.getVariable(_tid, getParam<SolverVariableName>("u")))),
+    _v_var(params.isParamValid("v")
+               ? dynamic_cast<const MooseLinearVariableFVReal *>(
+                     &_fe_problem.getVariable(_tid, getParam<SolverVariableName>("v")))
+               : nullptr),
+    _w_var(params.isParamValid("w")
+               ? dynamic_cast<const MooseLinearVariableFVReal *>(
+                     &_fe_problem.getVariable(_tid, getParam<SolverVariableName>("w")))
+               : nullptr)
+{
+  if (_use_nonorthogonal_correction || _use_deviatoric_terms)
+    _var.computeCellGradients();
+
+  Moose::FV::setInterpolationMethod(*this, _advected_interp_method, "advected_interp_method");
+
+  if (!_u_var)
+    paramError("u", "the u velocity must be a MooseLinearVariableFVReal.");
+
+  if (_dim >= 2 && !_v_var)
+    paramError("v",
+               "In two or more dimensions, the v velocity must be supplied and it must be a "
+               "MooseLinearVariableFVReal.");
+
+  if (_dim >= 3 && !_w_var)
+    paramError("w",
+               "In three-dimensions, the w velocity must be supplied and it must be a "
+               "MooseLinearVariableFVReal.");
+}
+
+void
+LinearWCNSFVMomentumFlux::initialSetup()
+{
+  // We need the cell gradients for two scenarios:
+  // (1) when we use nonorthogonal correction for the normal gradients
+  // (2) when we run with space-dependent viscosities where the deviatoric part of the
+  //     stress term can be non-negligible
+  if (_use_nonorthogonal_correction || !_mu.isConstant())
```

Comment:
relying on mu.isConstant is not very foolproof btw, because forwarding a constant through a GenericFunctorMaterial (commonly done) for example wont be caught
maybe use the _deviatoric_term boolean instead?